import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CMoveService.java

```diff
@@ -0,0 +1,240 @@
+
+    private CMoveTask(Association as, PresentationContext pc,
+        Attributes cmd, Attributes keys) {
+      super(as, pc, cmd);
+
+      this.keys = keys;
+    }
+
+    @Override
+    public void run() {
+      List<String> failedInstanceUids = new ArrayList<>();
+      ICStoreSender cstoreSender = null;
+      try {
+        if (canceled) {
+          throw new CancellationException();
+        }
+        runThread = Thread.currentThread();
+
+        AetDictionary.Aet cstoreTarget = aets.getAet(cmd.getString(Tag.MoveDestination));
+        if (cstoreTarget == null) {
+          sendErrorResponse(Status.MoveDestinationUnknown,
+              "Unknown AET: " + cmd.getString(Tag.MoveDestination));
+          return;
+        }
+
+        // need to get instances belonging to series/study
+        Attributes keysCopy = new Attributes(keys);
+        keysCopy.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+        String qidoPath;
+        try {
+          qidoPath = AttributesUtil.attributesToQidoPath(keysCopy,
+              TagUtils.toHexString(Tag.TransferSyntaxUID));
+          log.info("CMove QidoPath: " + qidoPath);
+        } catch (DicomServiceException e) {
+          log.error("CMove QidoPath error");
+          sendErrorResponse(e.getStatus(), e.getMessage(), e.getDataset(), null);
+          return;
+        }
+
+        JSONArray qidoResult;
+        try {
+          qidoResult = dicomWebClient.qidoRs(qidoPath);
+          if (qidoResult == null || qidoResult.length() == 0) {
+            throw new IDicomWebClient.DicomWebException("No instances to move");
+          }
+        } catch (IDicomWebClient.DicomWebException e) {
+          log.error("Failed QidoRs for CMove", e);
+          sendErrorResponse(Status.UnableToCalculateNumberOfMatches, e.getMessage());
+          return;
+        }
+
+        cstoreSender = cstoreSenderFactory.create();
+
+        int successfullInstances = 0;
+        int remainingInstances = qidoResult.length();
+        for (Object instance : qidoResult) {
+          sendPendingResponse(remainingInstances, successfullInstances, failedInstanceUids.size());
+
+          if (canceled) {
+            throw new CancellationException();
+          }
+
+          JSONObject instanceJson = (JSONObject) instance;
+          String studyUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.StudyInstanceUID));
+          String seriesUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SeriesInstanceUID));
+          String instanceUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SOPInstanceUID));
+          String classUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SOPClassUID));
+          String tsuid = AttributesUtil.getTagValueOrNull(instanceJson,
+              TagUtils.toHexString(Tag.TransferSyntaxUID));
+          if (tsuid == null) {
```

Comment:
Should we just throw an error if transfer syntax of the file is unknown? What if it's not ExplicitVRLittleEndian?

Reply:
DCM4che QIDO-RS didn't return it for some reason, this default worked. 
However, I've just run some tests against cloud healthcare api and while I didn't see it not return TransferSyntaxUID, returned TransferSyntaxUID was wrong in some cases (= did not match metadata in actual file transferred via stow-rs, which leads to silent failure whens client tried to download these).
Since QIDO-RS value proved to be untrustworthy, I've changed code to use TransferSyntaxUID from file transferred via stow-rs. Seems to work without problems now.

Reply:
Hmm, it should always be accurate in the CHC API since it's derived from the uploaded file. But the new way of doing it looks good to me.